File: mapgen.hs

import System.Random

import System.Console.GetOpt
import System.Environment(getArgs, getProgName)
import Control.Monad.State

-- Thanks to monoidal for this trick.
-- https://gist.github.com/2877633
type Rand = State StdGen
rand m = do a <- state next
            return (a `mod` m)

type Coord = (Int,Int)
type Range = (Int,Int)
type Area  = (Coord,Coord) -- Upper-left and lower-right bounds.

data Tile = TFloor | TWall
instance Show Tile where
  show TFloor = "."
  show TWall  = "#"

type MRow = [Tile]
type RMap = [MRow] 
  
showMap :: RMap -> String
showMap = unlines . map (>>= show)

makeMap :: Coord -> RMap
makeMap (x,y) = replicate y (replicate x TWall)

splitGap :: Int -> Int -> [a] -> ([a],[a],[a])
splitGap start size lst = (before, middle, after)
  where 
    (before,rest) = splitAt start lst
    (middle,after) = splitAt (abs size) rest

digRow :: Range -> MRow -> MRow
digRow (start,end) row = 
  before ++ replicate size TFloor ++ after
  where
    size = end - start + 1
    (before,_,after) = splitGap start size row

digRoom :: RMap -> Area -> RMap
digRoom rmap ((x,y),(u,v)) =
  ybefore ++ map (digRow (x,u)) rows ++ yend
  where 
    (ybefore,rows,yend) = splitGap y (v-y+1) rmap

randomRoom :: Coord -> Rand Area
randomRoom (w,h) = do
  (x,u) <- gen_range w
  (y,v) <- gen_range h
  return ((x,y),(u,v))
  where 
    minlen = 3
    gen_range m = do
        a <- rand (m-minlen-1)
        let a' = a + 1
            brange = m - a' - minlen
        b <- rand (max 1 brange)
        let b' = b + a' + minlen - 1
        return (a',b')
    
randomPoint :: Area -> Rand Coord
randomPoint ((x,y),(u,v)) = 
  do x' <- rand (u-x+1)
     y' <- rand (v-y+1)
     return (x' + x, y' + y)

digHallway :: RMap -> Area -> RMap
digHallway m ((x,y),(u,v)) = 
    -- Dig from (x,y) to (u,y) to (u,v).
    foldl digRoom m 
        [((u,min y v),(u,max y v)),((min x u,y),(max x u,y))]

digRandomHallways :: RMap -> [Area] -> Rand RMap
digRandomHallways m rooms
  | length rooms < 2 = return m
  | otherwise = 
    do n <- rand (length $ tail rooms)
       end1 <- randomPoint (head rooms)
       end2 <- randomPoint (tail rooms!!n)
       let m' = digHallway m (end1, end2)
       digRandomHallways m' (tail rooms)
    
splatter :: Int -> RMap -> Rand RMap
-- Splatter n random rooms onto m.
splatter n m = 
  do rooms <- replicateM n $ randomRoom (length (head m),length m)
     digRandomHallways (foldl digRoom m rooms) rooms
  where 
    center ((x,y),(u,v)) = ((x+u) `quot` 2, (y+v) `quot` 2)
    
data Options = Options {optRooms::Int,optDimensions::Coord}

defaults :: Options
defaults = Options {optRooms=5,optDimensions=(80,60)}

options = 
  [Option "n" ["rooms"] 
      (ReqArg (\s op-> return op{optRooms=read s::Int}) "ROOMS")
      "Number of rooms to dig.",
   Option "d" ["dimensions"]
      (ReqArg (\s op-> case reads s :: [(Coord,String)] of
                  ((dims,_):_) -> 
                    return op { optDimensions = dims }
                  _ -> error "Dimensions must be in format (width,height)") 
              "DIMENSIONS") 
      "Dimensions of map."]

main = do
  -- Parse command line.
  argv <- getArgs
  let (actions,noops,msgs) = getOpt RequireOrder options argv
  ops <- foldl (>>=) (return defaults) actions
  let Options { optRooms=rooms, optDimensions=dimensions } = ops
  
  gen <- newStdGen
  putStrLn . showMap $ evalState (splatter rooms $ makeMap dimensions) gen

File: mapgen.lua

FLOOR = string.byte('.')
WALL  = string.byte('#')

function Dungeon( w, h )
   local d = {width=w, height=h, grid={}}

   for j = 1, h do
      d.grid[j] = {}
      for i = 1, w do
         d.grid[j][i] = WALL
   end end
   
   return setmetatable( d, {
       __index = function( d, j )
                    return d.grid[j]
                 end,
       
       __tostring = function( self )
                       local s = ''
                       for _,row in pairs(self.grid) do
                           for _,c in pairs(row) do
                               s = s .. string.char(c)
                           end

                           s = s .. '\n'
                       end
                       return s
                    end,
    })
end

-- Delay the initialization of the dungeon as we don't know its dimensions.
dungeon = {} 

function random_range( min, max, minlen )
    min = math.random( min, max - minlen )
    max = math.random( min + minlen, max )
    return min, max
end

function random_room()
    local MINLEN = 4
    l, r = random_range( 2, dungeon.width-1,  MINLEN )
    u, d = random_range( 2, dungeon.height-1, MINLEN )
    return {left=l, right=r, up=u, down=d}
end

function dig( room )
    for j = room.up, room.down do for i = room.left, room.right do
        dungeon[j][i] = FLOOR
    end end 
end

function minmax(x,y) return math.min(x,y), math.max(x,y); end

function dig_hallway( a, b )
    -- Dig from (ax,ay) to (bx,ay) to (bx,by)
    for x = math.min(a.x,b.x), math.max(a.x,b.x) do dungeon[a.y][x] = FLOOR; end
    for y = math.min(a.y,b.y), math.max(a.y,b.y) do dungeon[y][b.x] = FLOOR; end
end

function random_point( room )
    return { x = math.random(room.left,room.right),
             y = math.random(room.up,room.down) }
end

function splatter( n )
    local rooms = {}
    for i = 1, n do rooms[i] = random_room(); end

    for i = 1, n do
        dig( rooms[i] )
        if i < n then
            dig_hallway (
                random_point( rooms[i] ),
                random_point( rooms[math.random(i+1,n)] )
            )
        end 
    end  
end

math.randomseed( os.time() )
dungeon = Dungeon( 20, 20 )
splatter( 5 )
print( dungeon ) 

